pkg/controller/ippool/pool_controller.go

```diff
@@ -0,0 +1,332 @@
+	}
+	apiAvailable := apiserver != nil && apiserver.Status.State == v1.TigeraStatusReady
+
+	// Get all IP pools in the cluster.
+	currentPools := &pcv1.IPPoolList{}
+	err = r.client.List(ctx, currentPools)
+	if err != nil && !errors.IsNotFound(err) {
+		r.status.SetDegraded(operator.ResourceReadError, "error querying IP pools", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+
+	// Create a lookup map of pools owned by this controller for easy access.
+	// This controller will ignore any IP pools that it itself did not create.
+	ourPools := map[string]pcv1.IPPool{}
+	for _, p := range currentPools.Items {
+		if hasOwner(&p, installation) {
+			// This pool is owned by the Installation object, so consider it ours.
+			ourPools[p.Spec.CIDR] = p
+		}
+	}
+
+	// For each pool that is desired, but doesn't exist, create it.
+	toCreate := []client.Object{}
+	for _, p := range installation.Spec.CalicoNetwork.IPPools {
+		if pool, ok := ourPools[p.CIDR]; !ok || !reflect.DeepEqual(pool, p) {
+			// Create a new pool, or update an existing one.
+			res, err := p.ToProjectCalicoV1()
+			if err != nil {
+				r.status.SetDegraded(operator.ResourceValidationError, "error handling IP pool", err, reqLogger)
+				return reconcile.Result{}, err
+			}
+			if apiAvailable {
+				v3res, err := v1ToV3(res)
+				if err != nil {
+					r.status.SetDegraded(operator.ResourceValidationError, "error handling IP pool", err, reqLogger)
+					return reconcile.Result{}, err
+				}
+				toCreate = append(toCreate, v3res)
+			} else {
+				toCreate = append(toCreate, res)
+			}
+		}
+	}
+
+	// Check existing pools owned by this controller that are no longer in
+	// the Installation resource.
+	toDelete := []client.Object{}
+	for cidr, pool := range ourPools {
+		reqLogger.WithValues("cidr", cidr).Info("Checking if pool is still valid")
+		found := false
+		for _, p := range installation.Spec.CalicoNetwork.IPPools {
+			if p.CIDR == cidr {
+				found = true
+				break
+			}
+		}
+		if !found {
+			// This pool needs to be deleted. We only ever send deletes via the API server,
+			// since deletion requires rather complex logic. If the API server isn't available,
+			// we'll instead just mark the pool as disabled temporarily.
+			reqLogger.WithValues("cidr", cidr).Info("Pool needs to be deleted")
+			if apiAvailable {
+				// v3 API is available - send a delete request.
+				v3res, err := v1ToV3(&pool)
+				if err != nil {
+					r.status.SetDegraded(operator.ResourceValidationError, "error handling IP pool", err, reqLogger)
+					return reconcile.Result{}, err
+				}
+				toDelete = append(toDelete, v3res)
+			} else {
+				// v3 API is not available. Just mark the pool as disabled so that new allocations
+				// don't come from this pool. We'll delete it once the API server is available.
+				pool.Spec.Disabled = true
+				toCreate = append(toCreate, &pool)
+			}
+		}
+	}
+
+	// Update pools in the cluster.
+	// TODO: This is bypassing a lot of validaiton! For example:
```

Comment:
Since this is `TODO` do you expect that we'll add the validation listed here at some point to the operator?
Because of this perhaps we should only update with the v3 api so that validation can be done and therefore require the apiserver being available.
That would block fixing IPPools without the apiserver but that is the same situation a user would have been in before the operator could manage IPPools.

Reply:
Yep, I've done this now - modifications to pools of any form can only go through the API server.

The only exception is creating new pools when there are none.